cmd/apiserver-watcher/gcp_handler.go

```diff
@@ -0,0 +1,199 @@
+//  The host process checks /run/cloud-routes/ for files $VIP.up and $VIP.down, the
+//  apiserver-watcher sets the apiserver VIP.down file on failure to avoid receiving
+//  traffic directed to the apiserver when it is not ready.
+//
+// Example rules using vip 10.0.0.2 for the apiserver. Note that the host in this case
+// has additional forwarded IPs 35.184.123.49 and 35.188.173.243
+//
+// -A PREROUTING -m comment --comment "gcp LB vip DNAT" -j gcp-vips
+// -A OUTPUT -m comment --comment "gcp LB vip DNAT for local clients" -j gcp-vips
+// -A gcp-vips -d 10.0.0.2/32 -j REDIRECT
+// -A gcp-vips -d 35.184.123.49/32 -j REDIRECT
+// -A gcp-vips -d 35.188.173.243/32 -j REDIRECT
+// -A gcp-vips-local -d 35.184.123.49/32 -j REDIRECT
+// -A gcp-vips-local -d 35.188.173.243/32 -j REDIRECT
+// -A gcp-vips-local -d 10.0.0.2/32 -j REDIRECT
+
+const (
+	// name of nat chain for iptables masquerade rules
+	gcpMasqChainName      = "gcp-vips"
+	gcpLocalMasqChainName = "gcp-vips-local"
+)
+
+type gcpHandler struct {
+	vip string
+	// GCP Loadbalancers provide the forwarded IPs in the metadata instance
+	forwardedIPs []string
+	ipt          *iptables.IPTables
+	stopCh       chan struct{}
+}
+
+var _ handler = &gcpHandler{}
+
+func newGCPHandler(vip string) (handler, error) {
+	// FIXME add IPv6 once GCP supports it
+	ipt, err := iptables.New(
+		iptables.IPFamily(iptables.ProtocolIPv4),
+		iptables.Timeout(iptablesTimeout),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &gcpHandler{
+		vip:    vip,
+		ipt:    ipt,
+		stopCh: make(chan struct{}),
+	}, nil
+}
+
+// onFailure: stop syncing iptables rules but don't flush all of them
+// keep the ones that are still being forwarded to this host.
+func (gcp *gcpHandler) onFailure() error {
+	// stop syncing iptables rules
+	select {
+	case gcp.stopCh <- struct{}{}:
+		glog.V(4).Info("stopping iptables sync loop")
+	default:
+		glog.V(4).Info("no sync iptables running")
+	}
+	// wait for iptables sync loop to finish
+	time.Sleep(iptablesTimeout * time.Second)
+
+	// indicate the host script that handles GCP LB vips
+	// to not forward traffic to the apiserver VIP
+	glog.Infof("apiserver down: create downfile: %s.down", gcp.vip)
+	if err := writeVipStateFile(gcp.vip, "down"); err != nil {
+		return fmt.Errorf("error writing state file for vip UP: %v", err)
+	}
+
+	glog.Info("apiserver down: deleting iptables apiserver rules for local traffic")
+	return gcp.ipt.DeleteIfExists("nat", gcpLocalMasqChainName, "--dst", gcp.vip, "-j", "REDIRECT")
+}
+
+// onSuccess: either start routes service, or remove down file
+func (gcp *gcpHandler) onSuccess() error {
+	// indicate the host script that handles GCP LB vips that
+	// it should not delete the apiserver VIP
+	glog.Infof("apiserver down: deleting downfile: %s.down", gcp.vip)
+	if err := removeVipStateFile(gcp.vip, "down"); err != nil {
+		return fmt.Errorf("error writing state file for vip UP: %v", err)
+	}
```

Comment:
@squeed on success we just remove the vip.down file, no need to use the vip.up since we install the vip iptables rules if it doesn't already exist